### PR TITLE
support pickle

### PR DESCRIPTION
**Pull request by wxchan**
A quick fix for #149 

But test_pickle in test_engine.py cannot pass yet, because continue train with model and loaded model get different result.

For better solution, we need an api to save model to model buffer like [save_raw in xgboost](https://github.com/dmlc/xgboost/blob/6f16f0ef5829965ba420ef6937617600d286820a/python-package/xgboost/core.py#L981) OR a function to deserialize dumped model.

Dataset.save_binary seems not working too, see test_basic.py line 15.

### CODE REVIEW

**Comment by guolinke** [comment]
@wxchan I just try save_binary, and it can work.
Did you know why them get different results?

**Comment by wxchan** [comment]
@guolinke did you use test_basic.py I changed to test save_binary? It sometimes cause segfault.

The saved model may lose some accuracy? I tried:
```
gbm1 = lgb.train(params, lgb_train, init_model=gbm)
gbm.save_model('lgb.model')
gbm2 = lgb.train(params, lgb_train, init_model='lgb.model')
```
The results of gbm1 and gbm2 are a little different too. (after 5 demical places)

**Comment by guolinke** [comment]
@wxchan I can save_binary on r-package branch.. we will merge to these changes to master. and you can try it out.
I think you can use something like ```fabs(a-b) < 1e-5``` to compare result.

**Comment by wxchan** [comment]
@guolinke although the differences are small, I hope it's not a bug.

**Comment by guolinke** [comment]
@wxchan 
when predict with init model, it uses double type. 
But for the inner training, it uses float type. 
So the result will be slight different. 
You also can try directly predict for some data, and compare them row by row.

**Comment by wxchan** [comment]
@guolinke save binary still fails with following code, anything I miss?

```
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
import lightgbm as lgb
X_train, _, y_train, _ = train_test_split(*load_breast_cancer(True), test_size=0.1)
train_data = lgb.Dataset(X_train, max_bin=255, label=y_train, free_raw_data=False)
train_data.save_binary('train.bin')
```

**Comment by guolinke** [comment]
@wxchan fixed, the reason is the   ```data_filename_```  is random pointer while loading from matrix..

**Comment by guolinke** [comment]
@wxchan , I change the most of text output of float type to double precision. I think the result predicted from model file should be same with the result predicted from original model now. You can test it again.

**Comment by henry0312** [comment]
ok, the precision problem seems to be solved.
